pkg/storage/segments/store.go

```diff
@@ -271,6 +273,128 @@ func (s *segmentStore) Delete(ctx context.Context, path paths.Path) (err error)
 	return s.pdb.Delete(ctx, path)
 }
 
+// Repair retrieves an at-risk segment and repairs and stores lost pieces on new nodes
+func (s *segmentStore) Repair(ctx context.Context, path paths.Path, lostPieces []int) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	//Read the segment's pointer's info from the PointerDB
+	pr, err := s.pdb.Get(ctx, path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if pr.GetType() != pb.Pointer_REMOTE {
+		return Error.New("Cannot repair inline segment")
+	}
+
+	seg := pr.GetRemote()
+	pid := client.PieceID(seg.GetPieceId())
+
+	// Get the list of remote pieces from the pointer
+	originalNodes, err := s.lookupNodes(ctx, seg)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// get the nodes list that needs to be excluded
+	var excludeNodeIDs []dht.NodeID
+	for _, v := range originalNodes {
+		if v != nil {
+			excludeNodeIDs = append(excludeNodeIDs, node.IDFromString(v.GetId()))
+		}
+	}
+
+	//remove all lost pieces from the list to have only healthy pieces
+	for j := range originalNodes {
+		for i := range lostPieces {
+			if j == lostPieces[i] {
+				originalNodes[j] = nil
+			}
+		}
+	}
+
+	// count the number of nil nodes thats needs to be repaired
+	totalNilNodes := 0
+	for i := range originalNodes {
+		if originalNodes[i] == nil {
+			totalNilNodes++
+			continue
+		}
+	}
+
+	//Request Overlay for n-h new storage nodes
+	newNodes, err := s.oc.Choose(ctx, overlay.Options{Amount: totalNilNodes, Space: 0, Excluded: excludeNodeIDs})
+	if err != nil {
+		return err
+	}
+
+	totalRepairCount := len(newNodes)
+
+	//make a repair nodes list just with new unique ids
+	repairNodesList := make([]*pb.Node, len(originalNodes))
+	for j, vr := range originalNodes {
+		// find the nil in the original node list
+		if vr == nil {
+			// replace the location with the newNode Node info
+			totalRepairCount--
+			repairNodesList[j] = newNodes[totalRepairCount]
+		}
+	}
+
+	es, err := makeErasureScheme(pr.GetRemote().GetRedundancy())
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	signedMessage, err := s.pdb.SignedMessage()
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// download the segment using the nodes just with healthy nodes
+	rr, err := s.ec.Get(ctx, originalNodes, es, pid, pr.GetSize(), signedMessage)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// get io.Reader from ranger
+	r, err := rr.Range(ctx, 0, rr.Size())
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(r)
+
+	// puts file to ecclient
+	exp, err := ptypes.Timestamp(pr.GetExpirationDate())
+	if err != nil {
+		return err
+	}
+
+	//@TODO-ASK check the expiration timer
```

Comment:
What is this TODO?

Reply:
put the comment as an action item to pass the expiration time and not create a new time. Took care of it.